### ISSUE TITLE
False positive VkGraphicsPipelineCreateInfo-layout and VkComputePipelineCreateInfo-layout warnings

### ISSUE THREAD

**Issue by thatcosmonaut** [KhronosGroup/Vulkan-ValidationLayers#9355]
**Environment:**
 - OS: Windows
 - GPU and driver version: AMD Radeon RX 7900 XTX, AMD proprietary driver 24.5.1 (LLPC)
 - SDK or header version if building from repo: 1.4.304.0
 - Options enabled (synchronization, best practices, etc.): Standard Preset

**Describe the Issue**

Hello, I maintain the new SDL GPU API. We have a suite of example usage code over [here](https://github.com/TheSpydog/SDL_gpu_examples). On the latest SDK I'm getting a validator error on some of our examples that never appeared before and it seems like a false positive to me. 

```
VUID-VkGraphicsPipelineCreateInfo-layout-07990(ERROR / SPEC): msgNum: -278545623 - Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-layout-07990 ] Object 0: handle = 0x4b7df1000000002f, name = CustomSampling.frag, type = VK_OBJECT_TYPE_SHADER_MODULE; Object 1: handle = 0x2b424a0000000034, type = VK_OBJECT_TYPE_PIPELINE_LAYOUT; | MessageID = 0xef65bb29 | vkCreateGraphicsPipelines(): pCreateInfos[0].pStages[1] SPIR-V (VK_SHADER_STAGE_FRAGMENT_BIT) uses descriptor [Set 2, Binding 0, variable "Texture"] of type VK_DESCRIPTOR_TYPE_STORAGE_IMAGE but expected VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER or VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE.
The Vulkan spec states: If a resource variable is declared in a shader, and the descriptor type is not VK_DESCRIPTOR_TYPE_MUTABLE_EXT, the corresponding descriptor set in layout must match the descriptor type (https://vulkan.lunarg.com/doc/view/1.4.304.0/windows/1.4-extensions/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-layout-07990)
    Objects: 2
        [0] 0x4b7df1000000002f, type: 15, name: CustomSampling.frag
        [1] 0x2b424a0000000034, type: 17, name: NULL
```

The application code is here: https://github.com/TheSpydog/SDL_gpu_examples/blob/main/Examples/CustomSampling.c
And the SPIRV is here: https://github.com/TheSpydog/SDL_gpu_examples/blob/main/Content/Shaders/Compiled/SPIRV/CustomSampling.frag.spv

When I use SPIRV-Cross to get GLSL output, I see:
```glsl
layout(set = 2, binding = 0) uniform texture2D Texture;
```

So this is clearly not a sampled image, but for some reason the validator thinks it is. I also confirmed that the descriptor set layout is using VK_DESCRIPTOR_TYPE_STORAGE_IMAGE.

A similar error appears in this example: https://github.com/TheSpydog/SDL_gpu_examples/blob/main/Examples/ToneMapping.c
With this shader: https://github.com/TheSpydog/SDL_gpu_examples/blob/main/Content/Shaders/Compiled/SPIRV/ToneMapReinhard.comp.spv

```
VUID-VkComputePipelineCreateInfo-layout-07990(ERROR / SPEC): msgNum: -2096418166 - Validation Error: [ VUID-VkComputePipelineCreateInfo-layout-07990 ] Object 0: handle = 0x612f93000000004e, type = VK_OBJECT_TYPE_SHADER_MODULE; Object 1: handle = 0x603cd90000000044, type = VK_OBJECT_TYPE_PIPELINE_LAYOUT; | MessageID = 0x830b328a | vkCreateComputePipelines(): pCreateInfos[0].stage SPIR-V (VK_SHADER_STAGE_COMPUTE_BIT) uses descriptor [Set 0, Binding 0, variable "InImage"] of type VK_DESCRIPTOR_TYPE_STORAGE_IMAGE but expected VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER or VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE.
The Vulkan spec states: If a resource variable is declared in a shader, and the descriptor type is not VK_DESCRIPTOR_TYPE_MUTABLE_EXT, the corresponding descriptor set in layout must match the descriptor type (https://vulkan.lunarg.com/doc/view/1.4.304.0/windows/1.4-extensions/vkspec.html#VUID-VkComputePipelineCreateInfo-layout-07990)
    Objects: 2
        [0] 0x612f93000000004e, type: 15, name: NULL
        [1] 0x603cd90000000044, type: 17, name: NULL
```

And similarly, when I use SPIRV-Cross...

```glsl
layout(set = 0, binding = 0) uniform texture2D inImage;
```

The examples render correctly which further leads me to believe that this is not an actual mistake.

**Expected behavior**

There should be no validator error.

**Comment by spencer-lunarg** [comment]
@thatcosmonaut how is this SPIR-V generated? Looking at the first example, I see it pointing to

> OpTypeImage %19 2D 2 0 0 1 Unknown

notice the `1` here, that means "sampled" here

![Image](https://github.com/user-attachments/assets/9843d619-2cbd-4c27-862c-ca328b25919f)

The Vulkan Spec also says this

![Image](https://github.com/user-attachments/assets/9a997c59-e35d-40c7-9478-54053525263d)

------

So this doesn't look like a bug in the Validation Layers, but who ever is generating this SPIR-V needs to set the `OpTypeImage` to have a `Sampled` of `2`

**Comment by thatcosmonaut** [comment]
We're using DXC to compile HLSL to SPIRV... let me look into this.

**Comment by spencer-lunarg** [comment]
ok, I am curious what the HLSL input was, trying SPIRV-Cross via HLSL, it doesn't seem to produce valid code https://godbolt.org/z/WfEnv4rzn

**Comment by spencer-lunarg** [comment]
@thatcosmonaut I think I got your answer - https://godbolt.org/z/KG3zxh3r3

```patch
- Texture2D<float4> Texture : register(t0, space2);
+ RWTexture2D<float4> Texture : register(u0, space2);
```

This makes the Image a "writable" image (aka a Storage Image)

Also found that some GPU don't care, while others have special things they do if they know the Image is Read Only

**Comment by thatcosmonaut** [comment]
Indeed, it does appear that DXC only compiles using Sampled = 2 if the resource is declared as RWTexture. 

Looks like the recommended approach is to force decorate with NonWritable and then enable SPVC_COMPILER_OPTION_HLSL_NONWRITABLE_UAV_TEXTURE_AS_SRV in SPIRV-Cross:

```hlsl
[[vk::ext_decorate(24)]] RWTexture2D<float4> Texture : register(u0, space2);
```

This horrific but it's my nightmare to deal with and not yours. Sorry for the false false positive!

**Comment by spencer-lunarg** [comment]
> This is horrific but it's my nightmare to deal with and not yours. Sorry for the false false positive!

Happy to figure it out and glad to know the layers are doing their job! Better here-and-now then on someone's random device with a random old GPU/Driver crashing

**Comment by thatcosmonaut** [comment]
Turns out that we can fix this on our end just by declaring our "readonly storage texture" descriptors as `VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE` instead of `VK_DESCRIPTOR_TYPE_STORAGE_IMAGE`. Works just fine.

**Comment by spencer-lunarg** [comment]
> Turns out that we can fix this on our end just by declaring our "readonly storage texture" descriptors as VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE instead of VK_DESCRIPTOR_TYPE_STORAGE_IMAGE. Works just fine.

Yes, also now re-looking, if you use `Texture2D` and try to write to it, DXC will complain for you

the `SAMPLED_IMAGE` and `STORAGE_IMAGE` main difference is you the Read vs Read/Write

Even with a Sampled Image you are currently doing a `OpImageFetch` which like an `OpImageRead` that they both don't use the sampler
